kitsune/wiki/utils.py

```diff
@@ -103,56 +108,68 @@ def _active_contributors_id(from_date, to_date, locale, product):
         WikiDocumentVisits.objects.filter(period=LAST_7_DAYS)
+        .select_related("document")
+        .prefetch_related(
+            "document__products",
+            "document__topics",
+            Prefetch(
+                "document__translations",
+                queryset=(
+                    Document.objects.visible(
+                        locale=locale,
+                        current_revision__is_approved=True,
+                        is_archived=False,
+                        is_template=False,
+                    )
+                    if locale != settings.WIKI_DEFAULT_LANGUAGE
+                    else None
+                ),
+            ),
+        )
         .filter(
             document__restrict_to_groups__isnull=True,
             document__locale=settings.WIKI_DEFAULT_LANGUAGE,
+            document__is_archived=False,
+            document__is_template=False,
         )
         .exclude(document__products__slug__in=settings.EXCLUDE_PRODUCT_SLUGS_FEATURED_ARTICLES)
-        .exclude(document__is_archived=True)
-        .exclude(document__is_template=True)
-        .order_by("-visits")
-        .select_related("document")
     )
 
+    # Add product and topics filters to the database query
     if product:
-        visits = visits.filter(document__products__in=[product.id])
+        visits = visits.filter(document__products=product)
+    if topics:
+        visits = visits.filter(document__topics__in=topics).distinct()
 
-    visits = visits[:10]
-    documents = []
+    # Order by visits but don't limit - we need enough documents for all topics
+    visits = visits.order_by("-visits")
+
+    # Convert to list to avoid additional queries
+    visits = list(visits)
 
+    # Get documents based on locale
+    documents = []
     if locale == settings.WIKI_DEFAULT_LANGUAGE:
-        for visit in visits:
-            documents.append(visit.document)
+        documents = [visit.document for visit in visits]
     else:
-        # prefretch localised documents to avoid n+1 problem
-        visits = visits.prefetch_related(
-            Prefetch(
-                "document__translations",
-                queryset=Document.objects.visible(
-                    locale=locale,
-                    current_revision__is_approved=True,
-                    is_archived=False,
-                    is_template=False,
-                ),
-            )
-        )
-
         for visit in visits:
-            translation = visit.document.translations.first()
-            if not translation:
-                continue
-            documents.append(translation)
+            translation = next(iter(visit.document.translations.all()), None)
+            if translation:
+                documents.append(translation)
 
-    if len(documents) <= 4:
-        return documents
-    return random.sample(documents, 4)
+    # Return all documents - the filtering per topic will happen in build_topics_data
+    return documents
```

Comment:
this no longer returns 4 articles like it used to. It will probably work for the topic cards but it will break functionality elsewhere

Reply:
You are right - this will break the homepage. I've reverted.